ecl/hqlcpp/hqlhtcpp.cpp

```diff
@@ -3562,8 +3546,28 @@ unsigned HqlCppTranslator::buildRtlField(StringBuffer * instanceName, IHqlExpres
             break;
         }
 
-        if (xpathName.length() && (xpathName.charAt(0) == '@'))
-            typeFlags |= RFTMhasxmlattr;
+        if (xpathName.length())
+        {
+            if (xpathName.charAt(0) == '@')
+                typeFlags |= RFTMhasxmlattr;
+            if (!strpbrk(xpathName, "/?*[]<>"))
+                typeFlags |= RFTMxpathscalar;
```

Comment:
Agreed that reversing the flag to "non-scalar" is better for backward compatibility, I'll change it.  Also agree that field flags would be better if it were compatible.  This does increase the number of type instances (I don't think by too much though).